File: extension/src/telemetry.js
/**
 * Copilot Telemetry Logger
 *
 * Records daily GitHub Copilot premium-interaction quota usage to a local
 * JSONL file inside ~/.activate/.  Based on ironarch-technology/copilot-telemetry-logger.
 *
 * Log schema (one JSON object per line):
 *   {
 *     "date": "2026-02-24",
 *     "timestamp": "2026-02-24T14:00:00.000Z",
 *     "premium_entitlement": 300,
 *     "premium_remaining": 142,
 *     "premium_used": 158,
 *     "quota_reset_date_utc": "2026-03-01T00:00:00Z",
 *     "source": "github_copilot_internal",
 *     "version": 1
 *   }
 */
const vscode = require('vscode');
const os = require('os');
const path = require('path');
const fs = require('node:fs/promises');

// ── Constants ───────────────────────────────────────────────────

const ACTIVATE_DIR = path.join(os.homedir(), '.activate');
const DEFAULT_LOG_FILE = 'copilot-telemetry.jsonl';
const NO_RESET_KEY = 'no-reset';
const FORCE_RESET_ENV = 'COPILOT_TELEMETRY_FORCE_RESET';

// Global-state keys (scoped to this module)
const LAST_RUN_KEY = 'telemetry.lastRunDate';
const LAST_FOCUS_RUN_TS_KEY = 'telemetry.lastFocusRunTs';
const CURRENT_QUOTA_RESET_KEY = 'telemetry.currentQuotaReset';

/** Minimum interval between focus-triggered log runs (1 hour). */
const FOCUS_THROTTLE_MS = 60 * 60 * 1000;

// ── Public API ──────────────────────────────────────────────────

/**
 * Initialise the telemetry logger.
 * Call once from the extension's `activate()` function.
 *
 * @param {vscode.ExtensionContext} context
 */
function initTelemetry(context) {
  // Register the manual command
  context.subscriptions.push(
    vscode.commands.registerCommand('activate-framework.telemetryRunNow', async () => {
      await runDailyLog(context, 'manual');
    }),
  );

  if (_isTestEnvironment()) return;
  if (!_isEnabled()) return;

  // Log on activation (startup)
  void runDailyLog(context, 'startup');

  // Re-log on window focus, throttled to once per hour
  const focusDisposable = vscode.window.onDidChangeWindowState((state) => {
    if (!state.focused) return;
    if (!_isEnabled()) return;
    const last = context.globalState.get(LAST_FOCUS_RUN_TS_KEY) ?? 0;
    if (shouldRunOnFocus(last, Date.now(), FOCUS_THROTTLE_MS)) {
      void runDailyLog(context, 'focus');
      void context.globalState.update(LAST_FOCUS_RUN_TS_KEY, Date.now());
    }
  });
  context.subscriptions.push(focusDisposable);
}

/**
 * Run the daily log (fetch quota → append entry).
 *
 * @param {vscode.ExtensionContext} context
 * @param {'startup'|'manual'|'focus'} trigger
 */
async function runDailyLog(context, trigger) {
  try {
    if (!_isEnabled() && trigger !== 'manual') return;

    const today = new Date();
    const todayKey = formatDateKey(today);
    const lastRunKey = context.globalState.get(LAST_RUN_KEY);

    if (trigger === 'startup' && lastRunKey === todayKey) return;

    const data = await fetchCopilotUserData();

    // Support forced quota-reset for testing
    const forcedReset = process.env[FORCE_RESET_ENV];
    if (forcedReset && forcedReset.trim().length > 0) {
      data.quota_reset_date_utc = forcedReset.trim();
    }

    const quota = extractPremiumQuota(data);

    const entry = {
      date: todayKey,
      timestamp: new Date().toISOString(),
      premium_entitlement: quota?.entitlement ?? null,
      premium_remaining: quota?.remaining ?? null,
      premium_used:
        quota && quota.entitlement != null && quota.remaining != null
          ? Math.max(0, quota.entitlement - quota.remaining)
          : null,
      quota_reset_date_utc: data.quota_reset_date_utc ?? null,
      source: 'github_copilot_internal',
      version: 1,
    };

    await appendDailyLog(context, entry);
    await context.globalState.update(LAST_RUN_KEY, todayKey);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error('Copilot Telemetry Logger failed:', message);
    if (trigger === 'manual') {
      vscode.window.showWarningMessage(
        `Copilot Telemetry Logger failed to record usage: ${message}`,
      );
    }
  }
}

/**
 * Read all log entries from the active JSONL file.
 * Used by the Usage page in the control panel.
 *
 * @returns {Promise<Array<object>>}
 */
async function readLogEntries() {
  const dirPath = _resolveLogDir();
  const activeFilePath = path.join(dirPath, DEFAULT_LOG_FILE);
  try {
    const raw = await fs.readFile(activeFilePath, 'utf8');
    return raw
      .split('\n')
      .filter((line) => line.trim().length > 0)
      .map((line) => {
        try {
          return JSON.parse(line);
        } catch {
          return null;
        }
      })
      .filter(Boolean);
  } catch {
    return [];
  }
}

/**
 * Get the resolved path to the active log file.
 * @returns {string}
 */
function getLogFilePath() {
  return path.join(_resolveLogDir(), DEFAULT_LOG_FILE);
}

// ── Internal helpers ────────────────────────────────────────────

/**
 * Check whether telemetry logging is enabled in settings.
 * @returns {boolean}
 */
function _isEnabled() {
  const config = vscode.workspace.getConfiguration('peregrine-activate');
  return config.get('telemetry.enabled', true);
}

/**
 * Resolve the directory for log files.
 * Uses `peregrine-activate.telemetry.logDirectory` setting, defaulting
 * to `~/.activate`.
 * @returns {string}
 */
function _resolveLogDir() {
  const config = vscode.workspace.getConfiguration('peregrine-activate');
  const configured = config.get('telemetry.logDirectory', '');
  const trimmed = configured.trim();

  if (trimmed.length === 0) return ACTIVATE_DIR;
  if (trimmed.startsWith('~')) return path.join(os.homedir(), trimmed.slice(1));
  return trimmed;
}

/**
 * Fetch GitHub Copilot internal user data including quota snapshots.
 * @returns {Promise<object>}
 */
async function fetchCopilotUserData() {
  const session = await vscode.authentication.getSession('github', ['user:email'], {
    createIfNone: true,
  });

  if (!session) {
    throw new Error('GitHub authentication session not available');
  }

  const response = await fetch('https://api.github.com/copilot_internal/user', {
    headers: {
      Authorization: `Bearer ${session.accessToken}`,
      Accept: 'application/json',
      'User-Agent': 'Peregrine-Activate-Telemetry',
    },
  });

  if (!response.ok) {
    throw new Error(`GitHub API returned ${response.status}: ${response.statusText}`);
  }

  const json = await response.json();
  return json ?? {};
}

/**
 * Extract the premium_interactions quota snapshot from the user data.
 * @param {object} data
 * @returns {{quota_id: string, entitlement?: number, remaining?: number}|null}
 */
function extractPremiumQuota(data) {
  const snapshots = data.quota_snapshots ? Object.values(data.quota_snapshots) : [];
  const premium = snapshots.find((q) => q.quota_id === 'premium_interactions');
  if (!premium || premium.unlimited) return null;
  return premium;
}

/**
 * Append a log entry to the active JSONL file, archiving first if the
 * quota reset date has changed.
 *
 * @param {vscode.ExtensionContext} context
 * @param {object} entry
 */
async function appendDailyLog(context, entry) {
  const dirPath = _resolveLogDir();
  await fs.mkdir(dirPath, { recursive: true });

  const currentQuotaKey = entry.quota_reset_date_utc ?? NO_RESET_KEY;
  const prevQuotaKey = context.globalState.get(CURRENT_QUOTA_RESET_KEY) ?? NO_RESET_KEY;

  await archiveActiveLogIfNeeded(dirPath, prevQuotaKey, currentQuotaKey, new Date());

  const activeFilePath = path.join(dirPath, DEFAULT_LOG_FILE);
  const line = JSON.stringify(entry) + '\n';
  await fs.appendFile(activeFilePath, line, 'utf8');

  await context.globalState.update(CURRENT_QUOTA_RESET_KEY, currentQuotaKey);
}

// ── Pure / testable helpers ─────────────────────────────────────

/**
 * Format a Date to `YYYY-MM-DD` (UTC).
 * @param {Date} date
 * @returns {string}
 */
function formatDateKey(date) {
  const year = date.getUTCFullYear();
  const month = String(date.getUTCMonth() + 1).padStart(2, '0');
  const day = String(date.getUTCDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
}

/**
 * Derive the archive date-stamp from the previous quota reset key.
 * @param {string} prevQuotaKey
 * @param {Date} now
 * @returns {string}
 */
function getArchiveDateStamp(prevQuotaKey, now) {
  if (prevQuotaKey !== NO_RESET_KEY) {
    const parsed = new Date(prevQuotaKey);
    if (!isNaN(parsed.getTime())) return formatDateKey(parsed);
  }
  return formatDateKey(now);
}

/**
 * Determine whether a focus-triggered run should proceed.
 * @param {number} lastRunTs - Epoch ms of last focus run
 * @param {number} nowMs     - Current epoch ms
 * @param {number} throttleMs
 * @returns {boolean}
 */
function shouldRunOnFocus(lastRunTs, nowMs, throttleMs) {
  return nowMs - lastRunTs >= throttleMs;
}

/**
 * If the quota reset key changed, rename the active log to an
 * archive file containing the date stamp.
 *
 * @param {string} dirPath
 * @param {string} prevQuotaKey
 * @param {string} currentQuotaKey
 * @param {Date} now
 * @returns {Promise<string|null>} Archive path if archival occurred
 */
async function archiveActiveLogIfNeeded(dirPath, prevQuotaKey, currentQuotaKey, now) {
  if (prevQuotaKey === currentQuotaKey) return null;

  const activeFilePath = path.join(dirPath, DEFAULT_LOG_FILE);

  try {
    await fs.stat(activeFilePath);
  } catch {
    return null;
  }

  const dateStamp = getArchiveDateStamp(prevQuotaKey, now);
  const archiveName = `copilot-telemetry-${dateStamp}.jsonl`;
  const archivePath = path.join(dirPath, archiveName);

  try {
    await fs.rename(activeFilePath, archivePath);
  } catch {
    await fs.copyFile(activeFilePath, archivePath);
    await fs.unlink(activeFilePath);
  }

  return archivePath;
}

/** @returns {boolean} */
function _isTestEnvironment() {
  return Boolean(
    process.env.VSCODE_EXTENSION_TEST ||
    process.env.VSCODE_TEST ||
    process.env.NODE_ENV === 'test',
  );
}

module.exports = {
  // Public
  initTelemetry,
  runDailyLog,
  readLogEntries,
  getLogFilePath,
  // Testable helpers
  formatDateKey,
  getArchiveDateStamp,
  shouldRunOnFocus,
  archiveActiveLogIfNeeded,
  extractPremiumQuota,
  FOCUS_THROTTLE_MS,
};

File: cli/telemetry.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

const (
	telemetryLogFile     = "copilot-telemetry.jsonl"
	copilotUserEndpoint  = "https://api.github.com/copilot_internal/user"
	telemetryUserAgent   = "Peregrine-Activate-Telemetry"
)

// TelemetryEntry is a single quota log entry.
type TelemetryEntry struct {
	Date              string  `json:"date"`
	Timestamp         string  `json:"timestamp"`
	PremiumEntitlement *int   `json:"premium_entitlement"`
	PremiumRemaining   *int   `json:"premium_remaining"`
	PremiumUsed        *int   `json:"premium_used"`
	QuotaResetDateUTC  string `json:"quota_reset_date_utc,omitempty"`
	Source             string `json:"source"`
	Version            int    `json:"version"`
}

// IsTelemetryEnabled checks if telemetry is opt-in enabled.
// Returns false if not explicitly enabled (opt-in, not opt-out).
func IsTelemetryEnabled(cfg Config) bool {
	return cfg.TelemetryEnabled != nil && *cfg.TelemetryEnabled
}

// ResolveGitHubToken returns a GitHub token from env or gh CLI.
func ResolveGitHubToken() string {
	if token := os.Getenv("GITHUB_TOKEN"); token != "" {
		return token
	}
	out, err := exec.Command("gh", "auth", "token").Output()
	if err == nil {
		return strings.TrimSpace(string(out))
	}
	return ""
}

// FetchCopilotUserData fetches Copilot quota data from the GitHub API.
func FetchCopilotUserData(token string) (map[string]interface{}, error) {
	if token == "" {
		return nil, fmt.Errorf("no GitHub token available (set GITHUB_TOKEN or run 'gh auth login')")
	}

	req, err := http.NewRequest("GET", copilotUserEndpoint, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", telemetryUserAgent)

	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("GitHub API returned %d", resp.StatusCode)
	}

	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20)) // 1 MB max
	if err != nil {
		return nil, err
	}

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	return data, nil
}

// ExtractPremiumQuota pulls the premium_interactions quota from user data.
func ExtractPremiumQuota(data map[string]interface{}) (entitlement, remaining *int) {
	snapshots, ok := data["quota_snapshots"].(map[string]interface{})
	if !ok {
		return nil, nil
	}
	for _, v := range snapshots {
		snap, ok := v.(map[string]interface{})
		if !ok {
			continue
		}
		if snap["quota_id"] != "premium_interactions" {
			continue
		}
		if unlimited, ok := snap["unlimited"].(bool); ok && unlimited {
			return nil, nil
		}
		if ent, ok := snap["entitlement"].(float64); ok {
			e := int(ent)
			entitlement = &e
		}
		if rem, ok := snap["remaining"].(float64); ok {
			r := int(rem)
			remaining = &r
		}
		return entitlement, remaining
	}
	return nil, nil
}

// BuildTelemetryEntry creates a log entry from API data.
func BuildTelemetryEntry(data map[string]interface{}) TelemetryEntry {
	now := time.Now().UTC()
	entry := TelemetryEntry{
		Date:      now.Format("2006-01-02"),
		Timestamp: now.Format(time.RFC3339Nano),
		Source:    "github_copilot_internal",
		Version:   1,
	}

	entitlement, remaining := ExtractPremiumQuota(data)
	entry.PremiumEntitlement = entitlement
	entry.PremiumRemaining = remaining

	if entitlement != nil && remaining != nil {
		used := *entitlement - *remaining
		if used < 0 {
			used = 0
		}
		entry.PremiumUsed = &used
	}

	if resetDate, ok := data["quota_reset_date_utc"].(string); ok {
		entry.QuotaResetDateUTC = resetDate
	}

	return entry
}

// telemetryLogDir returns the telemetry log directory (~/.activate).
func telemetryLogDir() string {
	return storeBase()
}

// telemetryLogPath returns the full path to the active telemetry log.
func telemetryLogPath() string {
	return filepath.Join(telemetryLogDir(), telemetryLogFile)
}

// AppendTelemetryEntry appends an entry to the JSONL log file.
func AppendTelemetryEntry(entry TelemetryEntry) error {
	dir := telemetryLogDir()
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	path := filepath.Join(dir, telemetryLogFile)

	line, err := json.Marshal(entry)
	if err != nil {
		return err
	}

	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()

	_, err = f.Write(append(line, '\n'))
	return err
}

// ReadTelemetryLog reads all entries from the JSONL log.
func ReadTelemetryLog() ([]TelemetryEntry, error) {
	path := filepath.Join(telemetryLogDir(), telemetryLogFile)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, nil
	}

	var entries []TelemetryEntry
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var e TelemetryEntry
		if json.Unmarshal([]byte(line), &e) == nil {
			entries = append(entries, e)
		}
	}
	return entries, nil
}

// ArchiveLogIfNeeded archives the active log when the quota reset date changes.
func ArchiveLogIfNeeded(currentResetDate, previousResetDate string) (string, error) {
	if currentResetDate == previousResetDate || previousResetDate == "" {
		return "", nil
	}

	dir := telemetryLogDir()
	activePath := filepath.Join(dir, telemetryLogFile)
	if _, err := os.Stat(activePath); err != nil {
		return "", nil
	}

	dateStamp := previousResetDate
	if t, err := time.Parse(time.RFC3339, previousResetDate); err == nil {
		dateStamp = t.Format("2006-01-02")
	}

	archiveName := fmt.Sprintf("copilot-telemetry-%s.jsonl", dateStamp)
	archivePath := filepath.Join(dir, archiveName)

	if err := os.Rename(activePath, archivePath); err != nil {
		// Fallback: copy + delete
		data, readErr := os.ReadFile(activePath)
		if readErr != nil {
			return "", readErr
		}
		if writeErr := os.WriteFile(archivePath, data, 0644); writeErr != nil {
			return "", writeErr
		}
		os.Remove(activePath)
	}

	return archivePath, nil
}

// RunTelemetry performs a single telemetry log run.
// token can be provided explicitly (e.g. from VS Code extension) or
// resolved automatically from env/gh CLI.
func RunTelemetry(token string) (*TelemetryEntry, error) {
	if token == "" {
		token = ResolveGitHubToken()
	}

	data, err := FetchCopilotUserData(token)
	if err != nil {
		return nil, err
	}

	entry := BuildTelemetryEntry(data)
	if err := AppendTelemetryEntry(entry); err != nil {
		return nil, err
	}

	return &entry, nil
}
